lib/Dialect/FIRRTL/Transforms/InferDomains.cpp

```diff
@@ -0,0 +1,425 @@
+  static DomainSequence canonicalizeDomains(ArrayRef<Value> domains) {
+    DomainSequence canonical(domains.begin(), domains.end());
+
+    // Sort domains deterministically.  Use block arg order if all are block
+    // args.  Fallback to opaque pointer comparison otherwise.  Note: the opaque
+    // pointer comparison is only stable within a run of MLIR and should not be
+    // relied upon for determinism outside of a run.
+    llvm::sort(canonical, [](Value a, Value b) {
+      auto aArg = dyn_cast<BlockArgument>(a);
+      auto bArg = dyn_cast<BlockArgument>(b);
+      if (aArg && bArg)
+        return aArg.getArgNumber() < bArg.getArgNumber();
+      return a.getAsOpaquePointer() < b.getAsOpaquePointer();
+    });
+
+    // Remove duplicates to handle cases like 'domains %A, %A'.
+    canonical.erase(std::unique(canonical.begin(), canonical.end()),
+                    canonical.end());
+
+    return canonical;
+  }
+
+public:
+  /// Get or create a domain variable for a value.
+  Value getDomainVar(Value value) {
+    auto it = valueToVar.find(value);
+    if (it != valueToVar.end())
+      return it->second;
+
+    // Create a new representative value for this domain variable.
+    Value representative = value;
+    valueToVar[value] = representative;
+    equivalenceClasses.insert(representative);
+    return representative;
+  }
+
+  /// Union two domain variables.  Returns success if successful, failure if
+  /// there's a domain conflict.  With canonical representation, conflicts only
+  /// occur when connecting values with truly different domain sets (not just
+  /// different ordering).
+  [[nodiscard]] LogicalResult unifyDomains(Value var1, Value var2) {
+    Value rep1 = equivalenceClasses.getLeaderValue(var1);
+    Value rep2 = equivalenceClasses.getLeaderValue(var2);
+
+    if (rep1 == rep2)
+      return success();
+
+    // Check for domain conflicts using canonical representation.
+    const DomainSequence &domains1 = getDomains(rep1);
+    const DomainSequence &domains2 = getDomains(rep2);
+
+    // Since domains are already canonical, we can compare them directly.
+    // Conflicts only occur when connecting truly different domain sets.  For
+    // example: 'domains %A' vs 'domains %B' is a conflict, but 'domains %A, %B'
+    // vs 'domains %B, %A' is not (same canonical form).
+    if (!domains1.empty() && !domains2.empty() && domains1 != domains2) {
+      // Conflict: both have different concrete domain sequences
+      return failure();
+    }
+
+    // Merge the concrete domain sequences (both are already canonical)
+    const DomainSequence &mergedDomains =
+        !domains1.empty() ? domains1 : domains2;
+
+    // Union the equivalence classes
+    equivalenceClasses.unionSets(rep1, rep2);
+
+    // Set the merged domain sequence on the new representative
+    if (!mergedDomains.empty()) {
+      Value newRep = equivalenceClasses.getLeaderValue(rep1);
+      concreteDomains[newRep] = mergedDomains;
```

Comment:
I think this has a lifetime issue. `mergedDomains` points to a reference to concreteDomains but concreteDomains[newRep] could grow capacity and invalidate the `mergedDomains`. You might want to change the type `concreateDomains` to `DenseMap<Value, std::uninque_ptr<SmallVector<..>>>`.